File: internal/limiter/sliding_window.go
package limiter

import (
	"context"
	"fmt"
	"time"

	aredis "aegis/internal/redis"
)

type SlidingWindow struct {
	client     *aredis.Client
	windowMs   int64
	limit      int64
	ttlSeconds int64
	keyPrefix  string
}

func NewSlidingWindow(client *aredis.Client, window time.Duration, limit, ttlSeconds int64) *SlidingWindow {
	return &SlidingWindow{
		client:     client,
		windowMs:   window.Milliseconds(),
		limit:      limit,
		ttlSeconds: ttlSeconds,
		keyPrefix:  "ratelimit:sliding_window",
	}
}

func (s *SlidingWindow) Name() string { return "sliding_window" }

func (s *SlidingWindow) Check(ctx context.Context, tenant, key string, requested int64) (Result, error) {
	fullKey := fmt.Sprintf("%s:%s:%s", s.keyPrefix, tenant, key)
	now := time.Now().UnixMilli()

	res, err := s.client.EvalShaOrLoad(ctx, "sliding_window", SlidingWindowScript,
		[]string{fullKey}, s.windowMs, s.limit, now, requested, s.ttlSeconds)
	if err != nil {
		return Result{}, fmt.Errorf("sliding_window check: %w", err)
	}

	vals, ok := res.([]interface{})
	if !ok || len(vals) != 4 {
		return Result{}, fmt.Errorf("sliding_window: unexpected script result: %#v", res)
	}

	allowed := toInt64(vals[0]) == 1
	remaining := toInt64(vals[1])
	resetTime := toInt64(vals[2])
	limit := toInt64(vals[3])

	var retryAfter int64
	if !allowed {
		retryAfter = resetTime - now
		if retryAfter < 0 {
			retryAfter = 0
		}
	}

	return Result{
		Allowed:    allowed,
		Remaining:  remaining,
		ResetTime:  resetTime,
		Limit:      limit,
		RetryAfter: retryAfter,
	}, nil
}

File: internal/limiter/lua/sliding_window.lua
-- sliding_window.lua
-- KEYS[1] = sorted-set key, e.g. ratelimit:sliding_window:{tenant}:{identifier}
-- ARGV[1] = window_ms (window size in milliseconds)
-- ARGV[2] = limit (max requests allowed within window)
-- ARGV[3] = now (milliseconds)
-- ARGV[4] = requested (number of "slots" this call wants to claim, usually 1)
-- ARGV[5] = ttl_seconds (key expiration)
--
-- Returns: { allowed(0/1), remaining, reset_time_ms, limit }
--
-- Implementation: sorted set where score = timestamp(ms), member = unique id
-- (timestamp:random) so concurrent requests at the same ms don't collide.
-- ZREMRANGEBYSCORE evicts anything older than (now - window), all inside one
-- EVAL so the count-then-add is atomic across concurrent nodes.

local key        = KEYS[1]
local window_ms  = tonumber(ARGV[1])
local limit      = tonumber(ARGV[2])
local now        = tonumber(ARGV[3])
local requested  = tonumber(ARGV[4]) or 1
local ttl_seconds = tonumber(ARGV[5])

local window_start = now - window_ms

-- Evict expired entries
redis.call("ZREMRANGEBYSCORE", key, "-inf", window_start)

local current = redis.call("ZCARD", key)

local allowed = 0
local remaining = math.max(0, limit - current)

if current + requested <= limit then
  for i = 1, requested do
    local member = tostring(now) .. ":" .. tostring(math.random(1, 1000000000)) .. ":" .. tostring(i)
    redis.call("ZADD", key, now, member)
  end
  allowed = 1
  remaining = math.max(0, limit - current - requested)
end

redis.call("EXPIRE", key, ttl_seconds)

-- reset_time: when the oldest entry currently in the window will fall out
local oldest = redis.call("ZRANGE", key, 0, 0, "WITHSCORES")
local reset_time = now + window_ms
if oldest and oldest[2] then
  reset_time = tonumber(oldest[2]) + window_ms
end

return { allowed, remaining, reset_time, limit }
